Refactor: Improve error messaging in sidebar YAML parser

Dedelweiss opened the pull request:
### Validations
On this PR I added some errors for different conditions : 

- [x] Error when a path in page is wrong 
- [x] Error when an index or a page path for a title is missing 
(It is therefore activated with the `title.isEmpty` condition. So far I've only found that when a user forgets a title it gives r = RawInput(,,,[],,false) in the toSidebar function. Thus, title.isEmpty is necessarily empty. However, if another error produces an r = RawInput(,,[],,false). This condition and its error message will no longer make sense. I am therefore less sure about this condition.)
- [x] Error when a title for a page is missing

- [x]  Warning when something is wrong but it does not interfere with the generation.

### Tests
I added two functions in SidebarParserTest.
The purpose of these functions is to check if it contains the `schemaMessage` which is the general template sent in all error messages + the unique error message for each error.

### IE
I also noticed that @gringrape wanted to work on this issue, if a code has already been done on his side, I am quite open to resume the work started or to close my PR to leave his. I am also open to any suggestions for adding errors.

Fixes: #15326

julienrf commented:
Hey @Dedelweiss, thank you very much for working on this!

About the first point: check the code that calls `Sidebar.toSidebar`, at some point you should get the root directory of the documentation files. Then you have two options: either pass the root directory to `Sidebar.toSidebar` and perform the check that the files exist here, or perform the check in the part of the code that calls `Sidebar.toSidebar` (that part of the code should have both the documentation root directory and the parsed sidebar). Does that make sense?

Dedelweiss commented:
> About the first point: check the code that calls `Sidebar.toSidebar`, at some point you should get the root directory of the documentation files. Then you have two options: either pass the root directory to `Sidebar.toSidebar` and perform the check that the files exist here

Hello @julienrf ! Thank you very much, it works perfectly.

Florian3k commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -30,13 +31,26 @@ object Sidebar:
+      Sidebar.Category(Option.when(title.nonEmpty)(title), Option.when(index.nonEmpty)(index), subsection.asScala.map(toSidebar(_, content)).toList, Option.when(dir.nonEmpty)(dir))
     case RawInput(title, page, index, subsection, dir, hidden) =>
-      report.error(s"Error parsing YAML configuration file.\n$schemaMessage")
+      val errors = (title.isEmpty(), page.isEmpty(), index.isEmpty(), subsection.isEmpty(), dir.isEmpty())
+      errors match
+        case (true, _, _, _, _) =>
+          report.error(s"Title is not provided.\n$schemaMessage")
+        case (false, true, true, _, _) =>
+          report.error(s"Index or page path to at least one page is missing.\n$schemaMessage")
+        case _ => report.error(s"Error parsing YAML configuration file.\n$schemaMessage")
```
Using `match` for this case seems like an overkill. Also subsection and dir are used at all.

I would suggest something like this:
```scala
if title.isEmpty() then
  ???
else if page.isEmpty() || index.isEmpty() then
  ???
else
  ???
```
or maybe even:
```scala
val msg =
  if title.isEmpty() then "Title is not provided."
  else if page.isEmpty() || index.isEmpty() then "Index or page path to at least one page is missing."
  else "Error parsing YAML configuration file."
report.error(s"$msg\n$schemaMessage")
```

Florian3k reviewed (approved):
Looks good, I left one comment.

julienrf commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -30,13 +31,26 @@ object Sidebar:
+          report.error(s"Title is not provided.\n$schemaMessage")
+        case (false, true, true, _, _) =>
+          report.error(s"Index or page path to at least one page is missing.\n$schemaMessage")
+        case _ => report.error(s"Error parsing YAML configuration file.\n$schemaMessage")
```
It seems the last line `case _ => report.error(...)` will _always_ report an error, even if the YAML was correctly formatted.

Florian3k replied:
Valid situations are handled by the two `case`s above.

julienrf reviewed (commented):
Is there a possibility to create tests that cover this feature?

Florian3k commented:
Sorry, I just realized there are some problems with validation logic. I will get back to you with more precise information soon.

Florian3k reviewed (changes requested):
Sorry, I just realized there are some problems with validation logic. I will get back to you with more precise information soon.

Dedelweiss commented:
Yes it's me, I'm currently working on it, sorry.

Dedelweiss commented:
Hi, I'm working on the test and fixing the problem of the error message when the sidebar is correctly formatted. 
I have sent a first commit for this. In this commit I changed the last one to a warning. Because even the static site is generated, there is a problem in the generation. For example, when you put :
```
- title: myFoo
   page: page1
   page: page2
```
Now I'm not sure if I missed any errors and I'm testing this. 
For the test, I haven't finished the one where the title is missing. Because I don't get a single error message like :
`report.error(s"$msg\n$schemaMessage")`
But several errors and an exception like this:

```
Static site YAML configuration file should comply with the following description:
The root element of static site needs to be <subsection>
`title` and `directory` properties are ignored in root subsection.

<subsection>:
  title: <string> # optional - Default value is file name. Title can be also set using front-matter.
  index: <string> # optional - If not provided, default empty index template is generated.
  directory: <string> # optional - By default, directory name is title name in kebab case.
  subsection: # optional - If not provided, pages are loaded from the index directory
    - <subsection> | <page>
  # either index or subsection needs to be present
<page>:
  title: <string> # optional - Default value is file name. Title can be also set using front-matter.
  page: <string>
  hidden: <boolean> # optional - Default value is false.

For more information visit:
https://docs.scala-lang.org/scala3/guides/scaladoc/static-site.html: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `java.util.ArrayList<dotty.tools.scaladoc.site.Sidebar$RawInput>` from Object value (token `JsonToken.START_OBJECT`)
 at [Source: (StringReader); line: 3, column: 5] (through reference chain: dotty.tools.scaladoc.site.Sidebar$RawInput["subsection"])
        at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)
        at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1741)
        at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1515)
Error parsing YAML configuration file: Title is not provided.
Static site YAML configuration file should comply with the following description:
The root element of static site needs to be <subsection>
`title` and `directory` properties are ignored in root subsection.

<subsection>:
  title: <string> # optional - Default value is file name. Title can be also set using front-matter.
  index: <string> # optional - If not provided, default empty index template is generated.
  directory: <string> # optional - By default, directory name is title name in kebab case.
  subsection: # optional - If not provided, pages are loaded from the index directory
    - <subsection> | <page>
  # either index or subsection needs to be present
<page>:
  title: <string> # optional - Default value is file name. Title can be also set using front-matter.
  page: <string>
  hidden: <boolean> # optional - Default value is false.

For more information visit:
https://docs.scala-lang.org/scala3/guides/scaladoc/static-site.html
Root element is not a subsection.
Static site YAML configuration file should comply with the following description:
The root element of static site needs to be <subsection>
`title` and `directory` properties are ignored in root subsection.

<subsection>:
  title: <string> # optional - Default value is file name. Title can be also set using front-matter.
  index: <string> # optional - If not provided, default empty index template is generated.
  directory: <string> # optional - By default, directory name is title name in kebab case.
  subsection: # optional - If not provided, pages are loaded from the index directory
    - <subsection> | <page>
  # either index or subsection needs to be present
<page>:
  title: <string> # optional - Default value is file name. Title can be also set using front-matter.
  page: <string>
  hidden: <boolean> # optional - Default value is false.

For more information visit:
https://docs.scala-lang.org/scala3/guides/scaladoc/static-site.html
```
I think because the title is missing, it goes to .fold and then to case _ in the toSidebar match. So I'm looking for a way to get a single error message.

Dedelweiss commented on scaladoc/test/dotty/tools/scaladoc/site/SidebarParserTest.scala:
```diff
@@ -34,6 +38,63 @@ class SidebarParserTest:
+      |          - page: my-page6/my-page6/my-page6.md
+      """.stripMargin
+
+  private val msgNoTitle = "Error parsing YAML configuration file: 'title' is not provided."
```
The problem is that in such a case (when a title is missing).  It is not possible to work on `r: RawInput` (,,,[],,false), even before the toSidebar function is called.

julienrf replied:
How could we make the error message clearer? Would something like this work? “the field 'title' is missing in one of the entries of the sidebar”

Otherwise, is it possible to return more information at the place where we try to access that `title` field and then use that information when we generate the error message?

Dedelweiss replied:
Yes of course, so in my last commit, if a title is missing, I take the raw content and look at a `page` without a `title` before.

Florian3k commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -30,16 +31,31 @@ object Sidebar:
+      Sidebar.Category(Option.when(title.nonEmpty)(title), Option.when(index.nonEmpty)(index), subsection.asScala.map(toSidebar(_, content)).toList, Option.when(dir.nonEmpty)(dir))
     case RawInput(title, page, index, subsection, dir, hidden) =>
-      report.error(s"Error parsing YAML configuration file.\n$schemaMessage")
+      if title.isEmpty() then
```
Why are we checking this? `title` is optional

Dedelweiss replied:
So it's optional for `subsection`, `- page` but not for a simple `page` from what I have seen in my tests. Example:
```
  - title: My title
    page: my-page1.md
```
But I noticed that it works with an index before (it's just ignored), so I'm thinking of adding `&& index.isEmpty`.
```
- index: my-page2.md
   page: my-page1.md
```

Dedelweiss replied:
Ho and this one works too:
```
index: index.md
subsection:
  - page: my-page1.md
    page: my-page2.md
  - page: my-page3.md
```
But the pages rendered are my-page2.md & my-page3.md

Dedelweiss commented:
Hello,
I'm sorry if this PR takes a long time, I'm trying to get as many errors as possible. I just updated the description. 
For the last else, I removed the msg `The parsing seems not to have been done correctly.` so as to not confuse the user who expects to know why the parsing went wrong.
Now, I don't know if it's better to let an `else` in case I missed a possible error and the user still gets an error message.
If you know of any other possible errors or have any feedback I am very interested. 
Thank you very much.

julienrf commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -30,16 +32,43 @@ object Sidebar:
+    }
+    val lines = fileContent.split("\n")
+    lines.zipWithIndex
+    .find { case (line, i) => line.trim.startsWith("page:") && !lines(i - 1).contains("- title:") }
```
I think the validation should be done in two steps: first, we parse the content as YAML, and second, we validate the YAML content. Here, this code is navigating the YAML data based on its text representation, which is very fragile (fore instance, in your code you make assumptions about the order of the fields `page` and `title`). Would it be possible for `pageWithNoTitle` to take as a parameter a YAML data structure?

julienrf commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -30,16 +32,43 @@ object Sidebar:
     case RawInput(title, page, index, subsection, dir, hidden) =>
-      report.error(s"Error parsing YAML configuration file.\n$schemaMessage")
+      if title.isEmpty() && index.isEmpty() then
+        val page = pageWithNoTitle(content).trim()
```
Actually I don’t understand this part. Why do we call `pageWithNoTitle` instead of using the `page` field of the scrutinee `r`?

Dedelweiss replied:
Because in some cases (including the case where a title is missing), the Try(ReadValue) function in load() does not succeed but still passes into the box where I perform my validations. So r = RawInput(,,,[],,false) and I can't get anything out of it.

julienrf replied:
OK I see. I think in that case we should not try to manually re-parse the sidebar to find out the `page`. Ultimately, we should probably use a different approach for parsing (maybe with [scala-yaml](https://virtuslab.github.io/scala-yaml/)?) that would allow us to navigate in the YAML data structure, but that’s a different topic.

I suggest we provide the best error messages we can without re-parsing the YAML in such a *ad-hoc* way. In that case, if we can’t get the `page`, just report “`title` is not provided for some page”. I know this is not very satisfactory, but I think we should really not try to re-parse the YAML text…

Dedelweiss replied:
Indeed it would be much better, I am working on it, thank you very much!

julienrf commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -30,16 +32,43 @@ object Sidebar:
+      else
+        val msg = s"Error parsing YAML configuration file."
+        report.warning(s"$msg\n$schemaMessage")
       Sidebar.Page(None, page, hidden)
```
Why do we still return a `Sidebar` even though we reported errors?

Dedelweiss commented:
> How is it that you know the sidebar is not valid, but you don’t know what is invalid?

So, I think I've tested all the possible errors with a minimised code but I thought that the security of sending an error if a user didn't pass one of my conditions wouldn't be bad.

julienrf reviewed (approved):
Thank you Lucas!

ckipp01 commented on scaladoc/src/dotty/tools/scaladoc/site/SidebarParser.scala:
```diff
@@ -30,16 +32,31 @@ object Sidebar:
     case RawInput(title, page, index, subsection, dir, hidden) =>
-      report.error(s"Error parsing YAML configuration file.\n$schemaMessage")
+      if title.isEmpty() && index.isEmpty() then
+        val msg = "`title` property is missing for some page."
```
Since we have access to `page` up above here, can't we specify which page is actually missing the `title` instead of just saying "some page"?

julienrf replied:
From what I understood this is not possible, see https://github.com/lampepfl/dotty/pull/17229#discussion_r1191381613

ckipp01 reviewed (commented):
Just a small nit.

julienrf commented:
It seems the tests added in this PR fail on Windows, see https://github.com/lampepfl/dotty/actions/runs/5240593619/jobs/9461621127. @Dedelweiss, would you have time to look at the failures?
